Update license files

TedThemistokleous opened the pull request:
Updating license at the top of AMD developed files. Created a script to achieve this called license_stamper.py 

All this does is pull in what LICENSE says, and adds that with the proper delimiters on top of our files. Let me know if I should be adding additional filenames. This is based on using what files are tracked in git, then running through them to stamp things.

I've also included a git hook to keep the date constantly rolling found from: https://github.com/ROCmSoftwarePlatform/rocBLAS-internal/pull/1214/files#

I can remove the above if it leads to any issues.

pfultz2 commented on .githooks/pre-commit:
```diff
@@ -45,6 +45,18 @@ else
 fi
 
+[[ -z "$files" ]] && exit
+# Change the copyright date at the top of any text files
```
Does this apply to all files or just the ones that were changed in the commit?

TedThemistokleous replied:
Should be all of them. I grabbed this from RocmBLAS commit Umang showed me.

pfultz2 replied:
We should reuse the query from files from below at line 61 and then assign it to the `files` variable: `files=$(git diff-index --cached --name-only $against | grep -E '\.h$|\.hpp$|\.cpp$|\.cl$|\.h\.in$|\.hpp\.in$|\.cpp\.in$|\.py$')`, and then use the `$files` variable below instead.

pfultz2 commented on .githooks/pre-commit:
```diff
@@ -45,6 +45,18 @@ else
+    if [[ -e $file ]]; then
+        /usr/bin/perl -pi -e 'INIT { exit 1 if !-f $ARGV[0] || -B $ARGV[0]; $year = (localtime)[5] + 1900 }
+            s/^([*\/#\/"*[:space:]]*)Copyright\s+(?:\(C\)\s*)?(\d+)(?:\s*-\s*\d+)?\s(Advanced\s*Micro\s*Devices)/qq($1Copyright (C) $2@{[$year != $2 ? "-$year" : ""]} $3)/ie
+            if $. < 10' "$file" && git add -u "$file"
```
Will this add the files to the current commit? If copyrights are updated we want to have that as a seperate commit.

pfultz2 commented on license_stamper.py:
```diff
@@ -0,0 +1,128 @@
+
+
+if __name__ == "__main__":
+    main()
```
This script should go under the tools directory.

pfultz2 commented on license_stamper.py:
```diff
@@ -0,0 +1,128 @@
+    #        licence_file = "LICENSE"
+
+    bashCommand = "cat LICENSE"
+    process = subprocess.Popen(bashCommand.split(), stdout=subprocess.PIPE)
```
I think `open('LICENSE').read()` would be simpler.

pfultz2 commented on license_stamper.py:
```diff
@@ -0,0 +1,128 @@
+    if error == None:
+        #Get a list of all the files in our git repo
+        bashCommand = "git ls-files --exclude-standard"
+        process = subprocess.Popen(bashCommand.split(), stdout=subprocess.PIPE)
```
Use `subprocess.run(bashCommand, capture_output=True, check=True, shell=True)` instead.

pfultz2 commented on examples/nlp/python_bert_squad/run_onnx_squad.py:
```diff
@@ -1,3 +1,26 @@
+#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+#  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+#  THE SOFTWARE.
+###########################################
```
The copyright below should be removed.

pfultz2 commented:
We should have a script so we can run a check in the CI to ensure there is a license and the copyright is up to date based on the last time it was modified.

umangyadav commented:
`cpp_mnist/digit/txt` is binary file, It got caught with the script. Can you remove change for that ?

umangyadav commented on src/include/migraphx/half.hpp:
```diff
@@ -1,3 +1,26 @@
+*  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+*  THE SOFTWARE.
+*/
 /*=============================================================================
```
This one also has dual licenses.

umangyadav commented on src/py/backend/backend.py:
```diff
@@ -1,3 +1,26 @@
+#  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+#  THE SOFTWARE.
+###########################################
 # -------------------------------------------------------------------------
```
Dual Licenses

umangyadav commented on src/targets/gpu/device/include/migraphx/gpu/device/types.hpp:
```diff
@@ -1,3 +1,26 @@
+*  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+*  THE SOFTWARE.
+*/
 /*=============================================================================
```
Dual Licenses

umangyadav commented on doc/src/conf.py:
```diff
@@ -1,3 +1,26 @@
+#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+#  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+#  THE SOFTWARE.
+###########################################
```
Looks like this file is auto-generated.  I am not sure if we should add a license on this or not.

TedThemistokleous replied:
How do we want to handle auto gen'd files then? This sounds liek I should have another file with a list of thing not to stamp then

pfultz2 replied:
Its not completely autogenerated. It was initially generated but then we modified with out settings, It should still have a license.

TedThemistokleous replied:
What about rules for generated files then? Can we assume since "we" generated it, we should stamp it with our license?

TedThemistokleous commented:
> @TedThemistokleous #1183 got merged in which added a few extra files. When you merge/rebase with develop branch, make sure to add licenses on those files.

Do you have a list of which files to update?

TedThemistokleous commented:
Updated this with changes to the script and removed the githook for now I borrowed from rocmBLAS.

causten commented on src/include/migraphx/half.hpp:
```diff
@@ -1,9 +1,26 @@
-/*=============================================================================
-    Copyright (c) 2017 Paul Fultz II
-    half.hpp
-    Distributed under the Boost Software License, Version 1.0. (See accompanying
```
License changed from Boost to MIT?

causten commented on tools/roctx.py:
```diff
@@ -1,3 +1,26 @@
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#####################################################################################
 #!/usr/bin/env python3
```
License should come after this line

TedThemistokleous replied:
Easy fix.

causten commented on src/include/migraphx/type_traits.hpp:
```diff
@@ -1,9 +1,26 @@
-    type_traits.hpp
-    Distributed under the Boost Software License, Version 1.0. (See accompanying
-    file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
-==============================================================================*/
```
License changed?

causten replied:
> Looks like from the email license was erroneously added.
> 
> I'll regenerate and revert things with an updated script that correctly handles shebangs (#!) at the start of scripts too.
> 
> @causten do we want this to also stamp our bash files as well or does the current bucket of cpp/python suffice?

yes that resolves my comment.  Thanks for investigating!

causten commented on src/targets/gpu/device/include/migraphx/gpu/device/types.hpp:
```diff
@@ -1,9 +1,26 @@
-    types.hpp
-    Distributed under the Boost Software License, Version 1.0. (See accompanying
-    file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
-==============================================================================*/
```
License changed?

causten reviewed (changes requested):
I'd like to get some clarity on the files that were previously licensed under Boost.

TedThemistokleous commented:
> I'd like to get some clarity on the files that were previously licensed under Boost.

Doing a grep search off develop shows those three files are licensed. 

Looks like it was some files which were originally licensed by @pfultz2 

`tthemist@ixt-rack-186 ~/AMDMIGraphX (develop) $ grep -HnIR " Software License" * --exclude-dir=depend
src/include/migraphx/half.hpp:4:    Distributed under the Boost Software License, Version 1.0. (See accompanying
src/include/migraphx/type_traits.hpp:4:    Distributed under the Boost Software License, Version 1.0. (See accompanying
src/targets/gpu/device/include/migraphx/gpu/device/types.hpp:4:    Distributed under the Boost Software License, Version 1.0. (See accompanying`

TedThemistokleous commented:
Looks like from the email license was erroneously added. 

I'll regenerate and revert things with an updated script that correctly handles shebangs (#!) at the start of scripts too.

@causten do we want this to also stamp our bash files as well or does the current bucket of cpp/python suffice?

causten commented:
> Looks like from the email license was erroneously added.
> 
> I'll regenerate and revert things with an updated script that correctly handles shebangs (#!) at the start of scripts too.
> 
> @causten do we want this to also stamp our bash files as well or does the current bucket of cpp/python suffice?

All source code files need a  license.  Yeah .bsh/.sh files too

TedThemistokleous commented:
Adding in a check_stamped.py script that goes over everything in the repo and determines if it needs to be stamped + determines if there's things outstanding/we haven't accounted her.

@pfultz2 do we want to be also handling json, and .ipynb files as well? The rest of the things seem to be just git scripts, build time files, and binarys (.proto/.onnx/.doocker)

For clarity here's what I ahve in the script right now


`    # The filetypes we want to check for that are stamped
    # LICENSE is included here as it SHOULD have a liscence in it otherwise flag it as unstamped
    supported_file_types = (".cpp", ".hpp", ".h", ".py", ".txt", ".sh", ".bsh",
                            "LICENSE")

    #add general stuff we shouldn't stamp and any exceptions here
    unsupported_file_types = [
        ".onnx", ".pb", ".rst", ".jpg", ".jpeg", ".proto", ".md", ".clang",
        ".cmake", ".weight", ".ipynb", ".ini", "json", ".docker", ".git",
        ".rules", ".yml"
    ]

    #Specifics to ignore
    specificIgnores = ("digits.txt", "Dockerfile", "Jenkinsfile", "")`

pfultz2 commented:
> do we want to be also handling json, and .ipynb files as well?

I am not sure, thats a good question for our legal department, but I think we can start with extensions you have now and extend it in the future.

umangyadav commented on src/py/backend/backend_rep.py:
```diff
@@ -1,3 +1,26 @@
+#  THE SOFTWARE.
+###########################################
 # -------------------------------------------------------------------------
 # Copyright (c) Advanced Micro Device Inc. All rights reserved.
```
Can you remove bottom license here ?

causten commented:
> > do we want to be also handling json, and .ipynb files as well?
> 
> I am not sure, thats a good question for our legal department, but I think we can start with extensions you have now and extend it in the future.

JSON no, Jupyter Notebook's yes.

umangyadav commented:
> @TedThemistokleous #1183 got merged in which added a few extra files. When you merge/rebase with develop branch, make sure to add licenses on those files.

Can you rebase wrt develop and add those files ? You can take a look at that PR to find out new files that needs to be stamped.

TedThemistokleous commented:
> > @TedThemistokleous #1183 got merged in which added a few extra files. When you merge/rebase with develop branch, make sure to add licenses on those files.
> 
> Can you rebase wrt develop and add licenses in those files ? You can take a look at that PR to find out new files that needs to be stamped.

The latest script I uploaded would flag those files actually (tools/check_stampled.py)
I added a list of file types we support and don't support + some ignores incase we have some oddball exceptions (like digits.txt) . I can add those files once they get flagged as unstamped.

@causten okay, let me fix and change things for the notebook files. 

Take a look at tools/check_stamped.py which just grabs the file types we support

TedThemistokleous commented:
> > > do we want to be also handling json, and .ipynb files as well?
> > 
> > 
> > I am not sure, thats a good question for our legal department, but I think we can start with extensions you have now and extend it in the future.
> 
> JSON no, Jupyter Notebook's yes.

Just added proper support for our notebook files and updated the ones in the repo.

TedThemistokleous commented on tools/check_stamped.py:
```diff
@@ -0,0 +1,151 @@
+
+
+if __name__ == "__main__":
+    main()
```
https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/pull/1255

TedThemistokleous commented:
Rebased off current develop branch. Targeting the other PR with the CI check and check_script to this branch so we can capture everything once this gets in.
